evm/src/cpu/kernel/asm/curve/secp256k1/ecrecover.asm

```diff
@@ -62,70 +62,58 @@ ecrecover_valid_input:
-    PUSH 0x79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798 // x-coordinate of generator
-    // stack: Gx, X, ec_add_valid_points_secp, u2, Y, pubkey_to_addr, retdest
-    SWAP1
-    // stack: X, Gx, ec_add_valid_points_secp, u2, Y, pubkey_to_addr, retdest
-    PUSH 0x483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 // y-coordinate of generator
-    // stack: Gy, X, Gx, ec_add_valid_points_secp, u2, Y, pubkey_to_addr, retdest
-    SWAP1
-    // stack: X, Gy, Gx, ec_add_valid_points_secp, u2, Y, pubkey_to_addr, retdest
-    SWAP4
-    // stack: u2, Gy, Gx, ec_add_valid_points_secp, X, Y, pubkey_to_addr, retdest
-    SWAP2
-    // stack: Gx, Gy, u2, ec_add_valid_points_secp, X, Y, pubkey_to_addr, retdest
-    %jump(ec_mul_valid_point_secp)
+// Computes `a * G + b * Q` using GLV+precomputation, where `G` is the Secp256k1 generator and `Q` is a point on the curve.
+// Pseudo-code:
+// precompute_table(G) -- precomputation table for the combinations of `G, phi(G), Q, phi(Q)`.
+// let a0, a1 = glv_decompose(a)
+// let b0, b1 = glv_decompose(b)
+// return msm_with_precomputation([a0, a1, b0, b1], [G, phi(G), Q, phi(Q)]) -- phi is the Secp endomorphism.
+ecdsa_msm_with_glv:
+    %stack (a, b, Qx, Qy, retdest) -> (a, ecdsa_after_glv_a, b, Qx, Qy, retdest)
+    %jump(glv_decompose)
+ecdsa_after_glv_a:
+    %stack (a1neg, a0, a1, b, Qx, Qy, retdest) -> (b, ecdsa_after_glv_b, a1neg, a0, a1, Qx, Qy, retdest)
+    %jump(glv_decompose)
+ecdsa_after_glv_b:
+    %stack (b1neg, b0, b1, a1neg, a0, a1, Qx, Qy, retdest) -> (a1neg, b1neg, Qx, Qy, ecdsa_after_precompute, a0, a1, b0, b1, retdest)
+    %jump(precompute_table)
+ecdsa_after_precompute:
+    // stack: a0, a1, b0, b1, retdest
+    PUSH 0 PUSH 0 PUSH 129
+    // stack: i, accx, accy, a0, a1, b0, b1, retdest
+ecdsa_after_precompute_loop:
+    %stack (i, accx, accy, a0, a1, b0, b1, retdest) -> (i, b1, i, accx, accy, a0, a1, b0, b1, retdest)
+    SHR %and_const(1)
+    %stack (bit_b1, i, accx, accy, a0, a1, b0, b1, retdest) -> (i, b0, bit_b1, i, accx, accy, a0, a1, b0, b1, retdest)
+    SHR %and_const(1)
+    %stack (bit_b0, bit_b1, i, accx, accy, a0, a1, b0, b1, retdest) -> (i, a1, bit_b0, bit_b1, i, accx, accy, a0, a1, b0, b1, retdest)
+    SHR %and_const(1)
+    %stack (bit_a1, bit_b0, bit_b1, i, accx, accy, a0, a1, b0, b1, retdest) -> (i, a0, bit_a1, bit_b0, bit_b1, i, accx, accy, a0, a1, b0, b1, retdest)
+    SHR %and_const(1)
+    %mul_const(2) ADD %mul_const(2) ADD %mul_const(2) ADD
+    %stack (index, i, accx, accy, a0, a1, b0, b1, retdest) -> (index, index, i, accx, accy, a0, a1, b0, b1, retdest)
+    %mul_const(2) %add_const(1)
+    %mload_kernel(@SEGMENT_KERNEL_ECDSA_TABLE)
+    SWAP1 %mul_const(2)
+    %mload_kernel(@SEGMENT_KERNEL_ECDSA_TABLE)
+    %stack (Px, Py, i, accx, accy, a0, a1, b0, b1, retdest) -> (Px, Py, accx, accy, ecdsa_after_precompute_loop_contd, i, a0, a1, b0, b1, retdest)
+    %jump(ec_add_valid_points_secp)
```

Comment:
I think I said this elsewhere, but you know that the two input points are non-zero here, so the checks at the start of `ec_add_valid_points_secp` are unnecessary.

Reply:
`Px,Py` can be `0,0` here.

Reply:
Ah, yes, I see. I was thinking that all table entries should be non-zero, but of course an adversary could select `P = -G` etc.